File: core/src/com/qtech/bubbleblaster/common/crash/CrashReport.java
package com.qtech.bubbleblaster.common.crash;

import com.qtech.bubbleblaster.addon.loader.AddonContainer;
import com.qtech.bubbleblaster.addon.loader.AddonManager;
import com.qtech.bubbleblaster.common.References;
import org.apache.commons.lang.SystemUtils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import oshi.SystemInfo;
import oshi.hardware.HardwareAbstractionLayer;

import javax.annotation.ParametersAreNonnullByDefault;
import java.io.*;
import java.nio.charset.StandardCharsets;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.AbstractMap;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

@SuppressWarnings("unused")
@ParametersAreNonnullByDefault
public final class CrashReport extends CrashCategory {
    private final List<CrashCategory> categories = new ArrayList<>();

    public CrashReport(String details, CrashReport report) {
        super(details);

        throwable = addCrashReport(report).throwable;
    }

    public CrashReport(String details, @Nullable CrashReport report, Throwable t) {
        this(details, t);

        if (t instanceof ReportedException) this.throwable = addReportedException((ReportedException) t).throwable;
        if (report != null) addCrashReport(report);
    }

    public CrashReport(String details, Throwable t) {
        super(details, t);

        if (t instanceof ReportedException) {
            this.throwable = addReportedException((ReportedException) t).throwable;
        }
    }

    private CrashReport addCrashReport(CrashReport report) {
        CrashCategory cat = new CrashCategory(report.getDetails(), report.getThrowable());
        cat.entries.clear();
        cat.entries.addAll(report.entries);
        addCategory(cat);

        for (CrashCategory category : report.getCategories()) {
            addCategory(category);
        }

        return report;
    }

    private CrashReport addReportedException(ReportedException exception) {
        CrashReport crashReport = exception.getCrashReport();
        CrashReport crashReport1 = new CrashReport(crashReport.details, crashReport.throwable);
        crashReport1.categories.addAll(crashReport.categories.subList(0, crashReport.categories.size() - 1));
        crashReport1.entries.addAll(crashReport.entries);
        return addCrashReport(crashReport1);
    }

    @NotNull
    public Throwable getThrowable() {
        return throwable;
    }

    public void addCategory(CrashCategory crashCategory) {
        this.categories.add(crashCategory);
    }

    public List<CrashCategory> getCategories() {
        return Collections.unmodifiableList(categories);
    }

    private CrashReport getFinalForm() {
        CrashReport crashReport = new CrashReport(details, throwable);
        crashReport.categories.addAll(categories);
        crashReport.entries.addAll(entries);

        Runtime runtime = Runtime.getRuntime();
        SystemInfo systemInfo = new SystemInfo();
        HardwareAbstractionLayer hardwareAbstractionLayer = systemInfo.getHardware();

        CrashCategory category = new CrashCategory("System Details");
        category.add("OS", System.getProperty("os.name") + " " + System.getProperty("os.version"));
        category.add("CPU", hardwareAbstractionLayer.getProcessor().getProcessorIdentifier().getName());
        category.add("GPU", hardwareAbstractionLayer.getGraphicsCards().get(0).getName());
        category.add("Memory", runtime.totalMemory() - runtime.freeMemory() + "/" + runtime.totalMemory());
        category.add("Loaded Addons", AddonManager.getInstance().getContainers().stream()
                .map(AddonContainer::getAddonId)
                .collect(Collectors.joining(", ")));
        category.add("Constructed Addons", AddonManager.getInstance().getContainers().stream()
                .filter((container) -> container.getJavaAddon() != null)
                .map(AddonContainer::getAddonId)
                .collect(Collectors.joining(", ")));

        crashReport.addCategory(category);
        return crashReport;
    }

    public ReportedException getReportedException() {
        return new ReportedException(getFinalForm());
    }

    @Override
    public String toString() {
        String s1 = "// " + details + "\r\n";
        StringBuilder cs = new StringBuilder();
        StringBuilder sb = new StringBuilder();

        Runtime runtime = Runtime.getRuntime();
        SystemInfo systemInfo = new SystemInfo();
        HardwareAbstractionLayer hardwareAbstractionLayer = systemInfo.getHardware();

//        List<AbstractMap.SimpleEntry<String, String>> entries = this.entries;
//        entries.add(new AbstractMap.SimpleEntry<>("OS", System.getProperty("os.name")));
//        entries.add(new AbstractMap.SimpleEntry<>("Processor", hardwareAbstractionLayer.getProcessor().getProcessorIdentifier().getName()));
//        entries.add(new AbstractMap.SimpleEntry<>("GPU", hardwareAbstractionLayer.getGraphicsCards().get(0).getName()));
//        entries.add(new AbstractMap.SimpleEntry<>("Memory", runtime.totalMemory() - runtime.freeMemory() + "/" + runtime.totalMemory()));
//        entries.add(new AbstractMap.SimpleEntry<>("Loaded Addons", AddonManager.getInstance().getContainers().stream()
//                .map(AddonContainer::getAddonId)
//                .collect(Collectors.joining(", "))));
//        entries.add(new AbstractMap.SimpleEntry<>("Constructed Addons", AddonManager.getInstance().getContainers().stream()
//                .filter((container) -> container.getJavaAddon() != null)
//                .map(AddonContainer::getAddonId)
//                .collect(Collectors.joining(", "))));

        if (entries.size() > 0) {
            sb.append("Details:").append(System.lineSeparator());
            for (AbstractMap.SimpleEntry<String, String> entry : entries) {
                sb.append("  ").append(entry.getKey());
                sb.append(": ");
                sb.append(entry.getValue());
                sb.append("\r\n");
            }
        }

        for (CrashCategory category : categories) {
            cs.append(System.lineSeparator()).append("=------------------------------------------------------------------=");
            cs.append(SystemUtils.LINE_SEPARATOR).append(category.toString());
        }

        cs.append("=------------------------------------------------------------------=");

        StringWriter sw = new StringWriter();
        PrintWriter pw = new PrintWriter(sw);
        throwable.printStackTrace(pw);

        return ">>> C R A S H   R E P O R T <<<\r\n" + s1 + "\r\n" + sw + cs + "\r\n" + sb;
    }

    @SuppressWarnings("ResultOfMethodCallIgnored")
    public void writeToFile() throws IOException {
        LocalDateTime now = LocalDateTime.now();
        String fileName = "CrashReport_" + now.format(DateTimeFormatter.ofPattern("MM.dd.yyyy-HH.mm.ss")) + ".txt";
        if (!References.CRASH_REPORTS.exists()) {
            References.CRASH_REPORTS.mkdirs();
        }
        FileOutputStream fileOutputStream = new FileOutputStream(new File(References.CRASH_REPORTS, fileName));
        fileOutputStream.write(toString().getBytes(StandardCharsets.UTF_8));
        fileOutputStream.flush();
        fileOutputStream.close();
    }
}

File: src/main/kotlin/com/qtech/bubbles/common/crash/CrashReport.kt
package com.qtech.bubbles.common.crash

import com.qtech.bubbles.addon.loader.AddonContainer
import com.qtech.bubbles.addon.loader.AddonManager
import com.qtech.bubbles.common.References
import org.apache.commons.lang.SystemUtils
import oshi.SystemInfo
import java.io.*
import java.nio.charset.StandardCharsets
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.*
import java.util.stream.Collectors
import javax.annotation.ParametersAreNonnullByDefault

@ParametersAreNonnullByDefault
class CrashReport : CrashCategory {
    private val categories: MutableList<CrashCategory> = ArrayList()

    constructor(details: String?, report: CrashReport?, t: Throwable?) : this(details, t) {
        if (t is ReportedException) this.throwable = addReportedException(t).throwable
        report?.let { addCrashReport(it) }
    }

    constructor(details: String?, report: CrashReport) : super(details) {
        this.throwable = addCrashReport(report).throwable
    }

    constructor(details: String?, t: Throwable?) : super(details, t) {
        if (t is ReportedException) {
            this.throwable = addReportedException(t).throwable
        } else {
            this.throwable = t
        }
    }

    private fun addCrashReport(report: CrashReport): CrashReport {
        val cat = CrashCategory(report.details, report.throwable)
        cat.entries.clear()
        cat.entries.addAll(report.entries)
        addCategory(cat)
        for (category in report.getCategories()) {
            addCategory(category)
        }
        return report
    }

    private fun addReportedException(exception: ReportedException): CrashReport {
        val crashReport = exception.crashReport
        val crashReport1 = CrashReport(crashReport.details, crashReport.throwable)
        crashReport1.categories.addAll(crashReport.categories.subList(0, crashReport.categories.size - 1))
        crashReport1.entries.addAll(crashReport.entries)
        return addCrashReport(crashReport1)
    }

    override var throwable: Throwable?
        set(throwable) {
            super.throwable = throwable
            field = throwable
        }

    fun addCategory(crashCategory: CrashCategory) {
        categories.add(crashCategory)
    }

    @Suppress("MemberVisibilityCanBePrivate")
    fun getCategories(): List<CrashCategory> {
        return Collections.unmodifiableList(categories)
    }

    private val finalForm: CrashReport
        get() {
            val crashReport = CrashReport(details, throwable)
            crashReport.categories.addAll(categories)
            crashReport.entries.addAll(entries)
            val runtime = Runtime.getRuntime()
            val systemInfo = SystemInfo()
            val hardwareAbstractionLayer = systemInfo.hardware
            val category = CrashCategory("System Details")
            category.add("OS", System.getProperty("os.name") + " " + System.getProperty("os.version"))
            category.add("CPU", hardwareAbstractionLayer.processor.processorIdentifier.name)
            category.add("GPU", hardwareAbstractionLayer.graphicsCards[0].name)
            category.add("Memory", (runtime.totalMemory() - runtime.freeMemory()).toString() + "/" + runtime.totalMemory())
            category.add(
                "Loaded Addons", AddonManager.instance.containers.stream()
                    .map(AddonContainer::namespace)
                    .collect(Collectors.joining(", "))
            )
            category.add("Constructed Addons", AddonManager.instance.containers.stream()
                .filter { container -> container.instance != null }
                .map(AddonContainer::namespace)
                .collect(Collectors.joining(", ")))
            crashReport.addCategory(category)
            return crashReport
        }
    val reportedException: ReportedException
        get() = ReportedException(finalForm)

    override fun toString(): String {
        val s1 = "// $details\r\n"
        val cs = StringBuilder()
        val sb = StringBuilder()

        if (entries.size > 0) {
            sb.append("Details:").append(System.lineSeparator())
            for ((key, value) in entries) {
                sb.append("  ").append(key)
                sb.append(": ")
                sb.append(value)
                sb.append("\r\n")
            }
        }
        for (category in categories) {
            cs.append(System.lineSeparator()).append("=------------------------------------------------------------------=")
            cs.append(SystemUtils.LINE_SEPARATOR).append(category.toString())
        }
        cs.append("=------------------------------------------------------------------=")
        val sw = StringWriter()
        val pw = PrintWriter(sw)
        throwable?.printStackTrace(pw)
        return ">>> C R A S H   R E P O R T <<<\r\n$s1\r\n$sw$cs\r\n$sb"
    }

    @Throws(IOException::class)
    fun writeToFile() {
        val now = LocalDateTime.now()
        val fileName = "CrashReport_" + now.format(DateTimeFormatter.ofPattern("MM.dd.yyyy-HH.mm.ss")) + ".txt"
        if (!References.CRASH_REPORTS.exists()) {
            References.CRASH_REPORTS.mkdirs()
        }
        val fileOutputStream = FileOutputStream(File(References.CRASH_REPORTS, fileName))
        fileOutputStream.write(toString().toByteArray(StandardCharsets.UTF_8))
        fileOutputStream.flush()
        fileOutputStream.close()
    }
}